### PR TITLE
feat: new resiliency metric for CircuitBreakerState (last_value / gauge)

### PR DESCRIPTION
**Pull request by jake-engelberg**
## Existing:
`dapr_resiliency_loaded`: The number of resiliency policies loaded.
`dapr_resiliency_count`: The number of times a resiliency policy has been executed.
`dapr_resiliency_activations_total`: Number of times a resiliency policy has been activated in a building block after a failure or after a state change.

## Added:
`dapr_resiliency_cb_state`: A resiliency policy's current CircuitBreakerState state. 4 series are generated, one for each state, with the tag status being `[unknown, closed, half-open, open]`. The current state is 1, all other states are 0.

Example view in grafana of the PromQL `dapr_resiliency_cb_state{}`:
Screenshot 2024-09-23 at 2 28 12 PM

https://github.com/dapr/dapr/issues/7476

### CODE REVIEW

**Comment by yaron2** [comment]
cc @halspang

**Comment by jake-engelberg** [comment]
> cc @halspang

Meant to have this in draft, but also wanted to bring up this Q, please take a look:
https://github.com/dapr/dapr/issues/7476#issuecomment-2161832889

File: tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	"github.com/dapr/dapr/tests/integration/framework/client"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(defaultcircuitbreaker))
+}
+
+type defaultcircuitbreaker struct {
+	daprdClient    *daprd.Daprd
```
Comment (filintod) [inline tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:41]: run format or gofumpt

File: tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apps
+
+import (
+	"context"
+	"fmt"
+	// "io"
```
Comment (filintod) [inline tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:19]: remove if not using

File: pkg/diagnostics/resiliency_monitoring.go [hunk 3]
```diff
@@ -80,12 +93,40 @@ func (m *resiliencyMetrics) PolicyLoaded(resiliencyName, namespace string) {
...
+
+		// Record 4 metrics, one for each cb state, with the active state having a value of 1
+		if policy == CircuitBreakerPolicy {
+			for _, s := range cbStatuses {
+				//Use correct status tag
+				commonTags = append(commonTags[:len(commonTags)-1], s)
+				stateActive := int64(0)
+				if s == status {
+					stateActive = int64(1)
+				}
```
Comment (filintod) [inline pkg/diagnostics/resiliency_monitoring.go:121]: we could remove the overwriting of commonTags, and also maybe these two could be a constant (ie `stateActive` and `stateInactive`), and just use:
```
if s == status {
  diagUtils.WithTags(m.circuitbreakerState.Name(), append(commonTags, s),
					m.circuitbreakerState.M(stateActive)
else {
  diagUtils.WithTags(m.circuitbreakerState.Name(), append(commonTags, s),
					m.circuitbreakerState.M(stateInactive)
}
```
this is kind of in the hot path, so a bit of repetition does not hurt to reduce allocations.

check to make sure it is correct

File: pkg/diagnostics/resiliency_monitoring_test.go [hunk 4]
```diff
@@ -277,6 +283,10 @@ func TestResiliencyCountMonitoringCBStates(t *testing.T) {
 			require.NoError(t, err)
 			require.Len(t, rows, test.wantNumberOfRows)
 
+			rowsCbState, err2 := view.RetrieveData(resiliencyCBStateViewName)
```
Comment (antontroshin) [inline pkg/diagnostics/resiliency_monitoring_test.go:286]: Any reason not to reuse `err`?

File: tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	daprdServer *daprd.Daprd
+	callCount1  *atomic.Int32
+	callCount2  *atomic.Int32
+}
+
+func (d *defaultcircuitbreaker) Setup(t *testing.T) []framework.Option {
+	d.callCount1 = &atomic.Int32{}
+	d.callCount2 = &atomic.Int32{}
+
+	handler := http.NewServeMux()
```
Comment (antontroshin) [inline tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:49]: Please use `tests/integration/framework/process/http/app`, see example [here](https://github.com/dapr/dapr/blob/cde4cd2d5d2196ddc6c62a0c416c5bab5f5d6fd0/tests/integration/suite/daprd/jobs/grpc/http.go#L53)

File: tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		daprd.WithAppPort(srv.Port()),
+		daprd.WithResourceFiles(resiliency),
+		daprd.WithAppID("client"),
+		daprd.WithLogLevel("debug"),
+	)
+	d.daprdServer = daprd.New(t,
+		daprd.WithAppPort(srv.Port()),
+		daprd.WithAppID("server"),
+	)
+	// log.Print(d.daprdClient.Metrics(t, context.Background()))
```
Comment (antontroshin) [inline tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:94]: Debug leftover?

File: tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		require.NoError(t, err)
+		resp, err := client.HTTP(t).Do(req)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, resp.StatusCode)
+		require.NoError(t, err)
+		require.NoError(t, resp.Body.Close())
+
+		// assert cb execution,activation,and current state counts
+		mtc := d.daprdClient.Metrics(t, context.Background())
+		assert.InDelta(t, float64(4), mtc["dapr_resiliency_count|app_id:client|flow_direction:outbound|name:myresiliency|namespace:|policy:circuitbreaker|status:closed|target:app_server"], 0)
```
Comment (antontroshin) [inline tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:129]: You can disable this lint check that requires `InDelta` and revert to the integer or equals assert.
See example in `component-contrib` repo https://github.com/dapr/components-contrib/blob/release-1.14/.golangci.yml#L266

Comment (jake-engelberg) [reply tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:127]: Does that disable float compare for the entire repo? I see some usage of `//nolint:testifylint` on a per file, possibly per line basis, but I'm fine using the InDelta as I switched to.

Comment (antontroshin) [reply tests/integration/suite/daprd/resiliency/apps/defaultcircuitbreaker.go:127]: > Does that disable float compare for the entire repo?

Yes. It is indeed might be excessive. That's NIT anyway.

**Review by antontroshin — DISMISSED** [review]

**Review by antontroshin — APPROVED** [review]

**Review by artursouza — APPROVED** [review]

**Pull request merged**